[#2] 회원가입 기능 구현

msugo1 opened the pull request:
- 유저의 서비스 가입을 위한 회원가입 기능추가
- 유저마다 고유한 아이디를 보장하기 위해 아이디 중복체크 구현
- pom.xml 오타 수정
- 오븐 회원가입 페이지 링크 (https://ovenapp.io/view/UNIZBKZnIC7r2VZLH3NrDrsAIxipEPN8/Pbbsr)

Commit:
[#2] 회원가입 기능 구현

- 유저의 서비스 가입을 위한 회원가입 기능추가
- 유저마다 고유한 아이디를 보장하기 위해 아이디 중복체크 구현

Commit:
[#2] 회원가입 기능 구현

- 유저의 서비스 가입을 위한 회원가입 기능추가
- 유저마다 고유한 아이디를 보장하기 위해 아이디 중복체크 구현
- pom.xml 설정 오류 수정

Commit:
[#2] 회원가입 기능 구현

- pom.xml 파일 설정 오타 수정

f-lab-dev commented on src/main/java/me/soo/helloworld/config/SecurityConfig.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http.csrf().disable().headers().frameOptions().disable();
```
이 메소드들은 각각 어떤 것들을 설정하는걸까요?

msugo1 replied:
- `HttpSecurity`: 특정 http 요청에 관해 보안설정을 가능하게 해줍니다.

> 기본값으로는 모든 요청이 지정되어 있으며, `matchers` 옵션을 지정해 특정 URL에만 보안설정을 적용하게 할 수 있습니다. 

- `csrf`: `Cross Site Request Forgery` (웹 사이트의 취약점을 이용하여 사용자가 의도하지 않은 요청을 통해 공격하는 방식)을 막기 위해 `csrf 토큰`을 header 정보에 포함시켜 서버에 요청함으로써  `csrf 프로텍션`을 설정하는 역할입니다. 

- `frameOptions`: `clickjacking attack`(취약점을 지닌 페이지에 공격자가 속성을 바꿔치기 함으로써 사용자가 클릭할 때 페이지가 의도된 방식과 다르게 작동하게 하여 공격하는 방식) 을 방지하기 위해 사용되는 `header 옵션`입니다.

- 이외에도 `hasRole` 이나 `hasAnyRole`의 옵션을 통해 특정 역할이 부여된 사용자에게만 접근을 허용하거나 `anonymous`옵션을 부여해 익명사용자들이 이용하게 할 수 있으며, `authenticated` 옵션을 부여해 인증된 사용자만 접근을 허용하게 만들 수도 있습니다.

<br>

**Bcrypt**의 `password** encoder`를 사용하기 위해 spring security를 의존성에 추가했는데 그 뒤로부터 이전에 성공했던 controller 요청 테스트들이  실패하여 원인을 찾다가 spring security가 기본값으로 주는 `csrf`와 `frameOption`이 문제라는 것을 알게 되었습니다. 설정 내용을 찾아봤는데 내용들이 지금 단계에서는 제가 이해하기에 어려움이 있고 시간이 많이 소요될 것 같아 일단 비활성화를 시켜두었습니다. 

<br>

혹여나 추가로 알아야 할 내용이나 보안설정을 미리 해놓는 것이 좋다면 피드백주시면 바로 수정하겠습니다.

f-lab-dev replied:
비밀번호 암호화 하나를 위해 굳이 `spring security`까지 가져다 쓰지 않을 것 같은데 이걸 가져다 쓰지 않고도 비밀번호는 암호화할 수 있지 않을까요?

msugo1 replied:
비밀번호 암호화하려고 구글링해보니 여러 옵션이 있더라구요. 그중에서 `BCrypt`가 제일 안전한 옵션처럼 보여서 찾아보니 `Spring Security`가 필요하다고 나와있어서 무조건 필요한 걸로 생각하고 있었어요. 한 번 다시 알아보겠습니다!

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserController.java:
```diff
@@ -0,0 +1,32 @@
+        return new ResponseEntity<>(HttpStatus.CREATED);
+    }
+
+    @GetMapping("/idCheck")
```
시간이 좀 들겠지만 이것도 한번 읽어보시면 좋겠습니다~

https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

msugo1 replied:
읽어 보겠습니다. 감사합니다

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserController.java:
```diff
@@ -0,0 +1,32 @@
+    @PostMapping("/signup")
+    public ResponseEntity<Void> userSignUp(@Valid @RequestBody User user) {
+        userService.insertUser(user);
+        return new ResponseEntity<>(HttpStatus.CREATED);
```
이 메소드는 항상 동일한 값을 리턴합니다. 이렇게 `new`를 사용하면 요청이 올 때마다 객체를 생성할 것 같은데 어떻게 해결할 수 있을까요?

msugo1 replied:
`HttpResponse` 전용 상수클래스를 만들어서 피드백 반영했습니다. 감사합니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserController.java:
```diff
@@ -0,0 +1,32 @@
+
+import javax.validation.Valid;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/users")
```
이 세 어노테이션들은 각각 어떤 역할을 하는 어노테이션일까요?

msugo1 replied:
### @RestController
- `Http Response Body`가 생성되는 방식으로 view를 리턴하는 `@Controller`와는 다르게 `Json/Xml` 등의 타입으로 `HTTP응답을 직접 리턴`한다는 특징을 지니고 있습니다. (스프링 부트는 Json을 위한 `MessageConverter`를 기본적으로 가지고 있습니다.)
- `@ResponseBody`와 `@Controller`가 **결합**된 형태로 클래스 위에 애노테이션을 선언해두면 더이상 매번 `@ResponseBody`를 붙일 필요가 없습니다.

msugo1 replied:
### @RequestMapping
- Request로 들어오는 `URL`을 특정 controller 클래스나 메소드로 연결시켜 주는  `mapping` 역할을 합니다.

- `[value 생략가능 = ("/url)] 값이 주어진 경우 해당 URL로 들어오는 요청을 그 URL의 값을 갖고 있는 메소드에 mapping시켜 줍니다.

- `header` 값을 부여해 특정 헤더가 있거나 없는 요청을 처리할 수 있습니다.

- `param` 값을 부여해 특정 파라미터가 있거나 없는 요청을 처리할 수 있습니다.

- `consumers`를 통해 처리할 `Content-Type`을 `produces`를 통해 받을 응답을 설정할 수 있습니다. 

- path variable이 주어진 경우 (`ex: /url/{id}`) `@pathVariable` annotation과 연계해 메소드의 파라미터로 path variable을 전달할 수 있습니다.

- `@RequestMapping(method = METHODNAME)`은 `@(METHODNAME)Mapping`으로 간소화 됩니다.

**해당 annotation이 `class 레벨`에서 사용될 경우**
1. `method = (METHOD)` 값이 주어져 있는 경우 해당 클래스 안에 있는 메소드는 기본적으로 위의 METHOD에 해당하는 `HTTP 메소드`를 따르게 됩니다.
2. `("/url")` URL값이 지정되어 있을 경우 해당 클래스 안에 있는 메소드의 URL은 기본적으로 클래스 레벨에 있는 URL을 앞에 먼저 붙이게 됩니다. `ex: /classurl/methodurl`

msugo1 replied:
### @RequiredArgsConstructor
- 초기화 되지 않은 `final `필드나 `@NotNull`이 붙은 field에 대한 생성자를 자동으로 제공해 줍니다. 따라서 중복되는 코드 작성을 줄일 수 있고 혹시 수정이 필요한 경우에도 수정할 부분이 적어집니다.
- 생성자에서 활용하고 싶은 field들을 구분지어줄 수 있습니다. (위의 final, @NotNull을 활용)
- 생성자를 활용한 의존성 주입을 할 때 편의성을 위해 활용합니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserController.java:
```diff
@@ -0,0 +1,32 @@
+    private final UserService userService;
+
+    @PostMapping("/signup")
+    public ResponseEntity<Void> userSignUp(@Valid @RequestBody User user) {
```
`@RequestBody`는 무슨 의미일까요?

msugo1 replied:
### @RequestBody
- HTTP요청의 `Body`를 `자바 객체`로 변환주는 역할을 합니다.
- Body가 존재해야하기 때문에 `GET 요청`에서는 `사용 불가능`하며 Body에  있는 `Json`이나 `XML`같은 형태의 데이터를 자바 객체로 바꿀 때  `Message Converter`를 활용합니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/model/User.java:
```diff
@@ -0,0 +1,47 @@
+ * */
+
+@Getter @Setter
+@Builder @AllArgsConstructor
```
1. 여기서는 2개씩 묶어주신 이유가 있으실까요?
2. `@Builder`는 사용하신 이유가 무엇인가요?

msugo1 replied:
1. 저는 2개씩 나란히 쓰는게 더 깔끔하다고 생각했었는데 `annotation`이 두 개씩 딱 안떨어지는 상황에서는 이 소스코드를 보는 다른 개발자들이 어노테이션을 실수로 못보고 지나치는 경우도 있을거라는 생각을 못했네요. 피드백 반영하고 다음부터는 유지보수하는 개발자들의 관점에서 보도록 노력하겠습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/service/UserService.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.stereotype.Service;
+
+public interface UserService {
+    public void insertUser(User user);
```
`Service`는 비지니스 로직을 작성하는 레이어입니다. `insert`는 어떤 레이어 관점의 용어일까요?

msugo1 replied:
다시 생각해보니 insert는 DAO쪽에 어울리는 이름이네요. userSignUp나 userRegister으로 수정하겠습니다

f-lab-dev commented on src/test/java/me/soo/helloworld/model/UserSignUpValidationTest.java:
```diff
@@ -0,0 +1,249 @@
+    }
+
+    @Test
+    @DisplayName("null을 userId에 넣으면 validation 위반이 발생합니다.")
```
이런 상황을 명시해주신건 좋네요~

f-lab-dev commented on src/test/java/me/soo/helloworld/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @Test
+    public void duplicateIdCheckTest() throws Exception {
```
여기는 성공하는 케이스만 테스트를 하고있는데 실패하는 케이스는 없을까요?

msugo1 replied:
그 생각을 못했었네요. 추가했고 `ResponseEntity` 매번 새로운 객체 생성하는 부분 해결되면 PR다시 날리겠습니다.

Commit:
[#2] 피드백 반영

* Controller 계층
- URL관련 웹 API 디자인 지침 반영(/idCheck -> /idcheck)
- 응답시 매번 새로운 응답 코드 객체를 생성하는 문제 해결

* Service 계층
- 유저 회원가입 관련 메소드 이름 변경 (insertUser -> userSignUp)

* Test
- duplicateIdCheck 테스트 실패 케이스 추가 및 @DisplayName에 테스트 내용 명시
- serviceImplTest를 Mockito를 이용하여 단위 테스트로 변경

f-lab-dev commented on src/main/java/me/soo/helloworld/service/UserService.java:
```diff
@@ -1,9 +1,8 @@
 import me.soo.helloworld.model.User;
-import org.springframework.stereotype.Service;
 
 public interface UserService {
```
이 클래스를 보면 `~~Service`, `~~ServiceImpl` 이렇게 개발하고있는데 이렇게 추상화시키신 이유가 무엇일까요?

f-lab-dev commented on src/test/java/me/soo/helloworld/controller/UserControllerTest.java:
```diff
@@ -70,17 +71,28 @@ public void userSignUpController() throws Exception {
 
     @Test
-    public void duplicateIdCheckTest() throws Exception {
+    @DisplayName("이미 등록되어 있는 아이디일 경우 Http Status Code 409(Conflict)를 리턴합니다.")
```
👍
